timemachine/fe/rbfe.py

```diff
@@ -679,3 +685,114 @@ def run_complex(
+            "failed:",
+            " | ".join(
+                [
+                    f"{edge.mol_a_name} -> {edge.mol_b_name} (kJ/mol)",
+                    f"exp_ddg {edge.metadata['exp_ddg_kcal']:.2f}" if "exp_ddg_kcal" in edge.metadata else "",
+                    f"fep_ddg {edge.metadata['fep_ddg_kcal']:.2f} +- {edge.metadata['fep_ddg_err_kcal']:.2f}",
+                ]
+            ),
+        )
+        print(err)
+        traceback.print_exc()
+        return None
+
+    path = f"success_rbfe_result_{edge.mol_a_name}_{edge.mol_b_name}.pkl"
+    pkl_obj = (mol_a, mol_b, edge.metadata, smarts, core, solvent_res, solvent_top, complex_res, complex_top)
+    file_client.store(path, pickle.dumps(pkl_obj))
+
+    solvent_ddg = np.sum(solvent_res.all_dGs)
+    solvent_ddg_err = np.linalg.norm(solvent_res.all_errs)
+    complex_ddg = np.sum(complex_res.all_dGs)
+    complex_ddg_err = np.linalg.norm(complex_res.all_errs)
+
+    tm_ddg = complex_ddg - solvent_ddg
+    tm_err = np.linalg.norm([complex_ddg_err, solvent_ddg_err])
+
+    print(
+        "finished:",
+        " | ".join(
+            [
+                f"{edge.mol_a_name} -> {edge.mol_b_name} (kJ/mol)",
+                f"complex {complex_ddg:.2f} +- {complex_ddg_err:.2f}",
+                f"solvent {solvent_ddg:.2f} +- {solvent_ddg_err:.2f}",
+                f"tm_pred {tm_ddg:.2f} +- {tm_err:.2f}",
+                f"exp_ddg {edge.metadata['exp_ddg_kcal']:.2f}" if "exp_ddg_kcal" in edge.metadata else "",
+                f"fep_ddg {edge.metadata['fep_ddg_kcal']:.2f} +- {edge.metadata['fep_ddg_err_kcal']:.2f}"
+                if "fep_ddg_kcal" in edge.metadata and "fep_ddg_err_kcal" in edge.metadata
+                else "",
+            ]
+        ),
+    )
+
+    return path
+
+
+def run_edges_parallel(
+    ligands: Sequence[Chem.rdchem.Mol],
+    edges: Sequence[Edge],
+    ff: Forcefield,
+    protein: app.PDBFile,
+    n_frames: int,
+    n_gpus: int,
+    seed: int,
+    pool_client: Optional[AbstractClient] = None,
+    file_client: Optional[AbstractFileClient] = None,
+):
+    mols = {get_mol_name(mol): mol for mol in ligands}
+
+    pool_client = pool_client or CUDAPoolClient(n_gpus)
+    pool_client.verify()
+
+    file_client = file_client or FileClient()
+
+    # Ensure that all mol props (e.g. _Name) are included in pickles
+    # Without this get_mol_name(mol) will fail on roundtripped mol
+    Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.AllProps)
+
+    # NOTE: using a generator expression here ensures that no references are kept to future objects after we've
+    # retrieved their results
+    jobs = (
+        pool_client.submit(
+            run_edge_and_save_results,
+            edge,
+            mols,
+            ff,
+            protein,
+            n_frames,
+            seed + edge_idx,
+            file_client,
+        )
+        for edge_idx, edge in enumerate(edges)
+    )
+
+    paths = [job.result() for job in jobs]
```

Comment:
Would be nice to do
```suggestion
    paths = [job.result() for job in futures.as_completed(jobs)]
```
here instead, but this doesn't seem possible without reimplementing some `concurrent.futures` internals for our custom futures (e.g. for AWS Batch)